src/main/java/picard/illumina/ExtractBarcodesProgram.java

```diff
@@ -0,0 +1,199 @@
+package picard.illumina;
+
+import htsjdk.samtools.metrics.MetricsFile;
+import htsjdk.samtools.util.StringUtil;
+import htsjdk.samtools.util.Tuple;
+import org.broadinstitute.barclay.argparser.Argument;
+import picard.cmdline.CommandLineProgram;
+import picard.cmdline.StandardOptionDefinitions;
+import picard.illumina.parser.*;
+import picard.illumina.parser.readers.BclQualityEvaluationStrategy;
+import picard.util.IlluminaUtil;
+
+import java.io.File;
+import java.text.NumberFormat;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+
+public abstract class ExtractBarcodesProgram extends CommandLineProgram {
+    @Argument(doc = "The distance metric that should be used to compare the barcode-reads and the provided barcodes for finding the best and second-best assignments.")
+    public DistanceMetric DISTANCE_MODE = DistanceMetric.HAMMING;
+
+    @Argument(doc = "Maximum mismatches for a barcode to be considered a match.")
+    public int MAX_MISMATCHES = 1;
+
+    @Argument(doc = "Minimum difference between number of mismatches in the best and second best barcodes for a barcode to be considered a match.")
+    public int MIN_MISMATCH_DELTA = 1;
+
+    @Argument(doc = "Maximum allowable number of no-calls in a barcode read before it is considered unmatchable.")
+    public int MAX_NO_CALLS = 2;
+
+    @Argument(shortName = "Q", doc = "Minimum base quality. Any barcode bases falling below this quality will be considered a mismatch even if the bases match.")
+    public int MINIMUM_BASE_QUALITY = 0;
+
+    @Argument(doc = "The minimum quality (after transforming 0s to 1s) expected from reads.  If qualities are lower than this value, an error is thrown." +
+            "The default of 2 is what the Illumina's spec describes as the minimum, but in practice the value has been observed lower.")
+    public int MINIMUM_QUALITY = BclQualityEvaluationStrategy.ILLUMINA_ALLEGED_MINIMUM_QUALITY;
+
+    @Argument(doc = "Lane number. ", shortName = StandardOptionDefinitions.LANE_SHORT_NAME)
+    public Integer LANE;
+
+    @Argument(doc = ReadStructure.PARAMETER_DOC, shortName = "RS")
+    public String READ_STRUCTURE;
+
+    @Argument(shortName = "GZIP", doc = "Compress output FASTQ files using gzip and append a .gz extension to the file names.")
+    public boolean COMPRESS_OUTPUTS = false;
```

Comment:
Is this common to all programs?  I.e. when barcode-calling on the fly we don't write .barcode.txt files so maybe this only applies to ExtractIlluminaBarcodes?

Reply:
This is whether or not we compress the fastqs as well.